flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp

```diff
@@ -786,13 +797,59 @@ llvm::LogicalResult BroadcastAssignBufferization::matchAndRewrite(
   mlir::Value shape = hlfir::genShape(loc, builder, lhs);
   llvm::SmallVector<mlir::Value> extents =
       hlfir::getIndexExtents(loc, builder, shape);
-  hlfir::LoopNest loopNest =
-      hlfir::genLoopNest(loc, builder, extents, /*isUnordered=*/true,
-                         flangomp::shouldUseWorkshareLowering(assign));
-  builder.setInsertionPointToStart(loopNest.body);
-  auto arrayElement =
-      hlfir::getElementAt(loc, builder, lhs, loopNest.oneBasedIndices);
-  builder.create<hlfir::AssignOp>(loc, rhs, arrayElement);
+
+  bool isArrayRef =
+      mlir::isa<fir::SequenceType>(fir::unwrapRefType(lhs.getType()));
+  if (lhs.isSimplyContiguous() && extents.size() > 1 &&
```

Comment:
Isn't `lhs.isSimplyContiguous()` always true for the allocatable arrays?  What is the reason for adding `isAllocatableArray` method?

Reply:
Yes, `lhs.isSimplyContiguous()` is always true for allocatable arrays.

I've added `isAllocatableArray` because the code now handles only `!fir.ref<!fir.array>` and `!fir.box<!fir.heap<!fir.array>>` types.

I guess it should also handle `!fir.box<!fir.array>`, when those are contiguous. They may appear when using non-default lower bounds.

By handling these cases, do you think it would be safe to drop the `isAllocatableArray` check? Or are there other array types that also pass the `isSimplyContiguous` check?

Reply:
I believe `isSimplyContiguous` may also return true for `!fir.box<!fir.ptr<...`.  I do not think you need to explicitly handle all these different cases.  There is `hlfir::derefPointersAndAllocatables` call above.  If `lhs` is a box, you may use the box_addr operation to get the base address of the array, otherwise, `lhs` is the base address of the array already.

Reply:
Done. Removing the `isAllocatableArray` check also made it easier to support `!fir.box<!fir.array>`.
Now all boxed arrays are converted to `!fir.box<!fir.array<?x type>>`